src/key.rs

```diff
@@ -1,43 +1,104 @@
-#[derive(Debug, Copy, Clone)]
+/// The `index` is specified by the user and is used to differentiate between multiple arroy indexes.
+/// Its last bit represent the mode we're in:
+/// - 0 means we're manipulating an item node.
+/// - 1 means we're manipulating a tree node.
+/// The metadata are stored at the last tree node.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub struct Key {
-    /// The prefix specified by the user.
-    pub index: u16,
-    pub node: NodeId,
-    /// Unused space.
-    _padding: u8,
+    // The prefix specified by the user on the first 7 bits
+    // And the last bit for the mode.
+    // - 0 => Item
+    // - 1 => Tree
+    pub index: u8,
+    pub item: ItemId,
+}
+
+impl fmt::Debug for Key {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        #[derive(Debug)]
+        enum Mode {
+            Item,
+            Tree,
+            Metadata,
+        }
+        let mode = match self.index & 1 {
+            0 => Mode::Item,
+            1 if self.item == ItemId::MAX => Mode::Metadata,
+            1 => Mode::Tree,
+            _ => unreachable!(),
+        };
+
+        f.debug_struct("Key")
+            .field("index", &(self.index & 0b1111_1110))
+            .field("mode", &mode)
+            .field("item", &self.item)
+            .finish()
+    }
 }
 
 impl Key {
-    pub const fn new(index: u16, node: NodeId) -> Self {
-        Self { index, node, _padding: 0 }
+    const ITEM_MASK: u8 = 0b1111_1110;
+    const TREE_MASK: u8 = !Self::ITEM_MASK;
+
+    pub const fn new(index: u8, item: ItemId) -> Self {
+        Self { index, item }
+    }
+
+    pub const fn metadata(index: u8) -> Self {
+        Self::new(index | Self::TREE_MASK, ItemId::MAX)
     }
 
-    pub const fn metadata(index: u16) -> Self {
-        Self::new(index, NodeId::metadata())
+    pub const fn item(index: u8, item: ItemId) -> Self {
+        Self::new(index & Self::ITEM_MASK, item)
     }
 
-    pub const fn item(index: u16, item: u32) -> Self {
-        Self::new(index, NodeId::item(item))
+    pub const fn tree(index: u8, item: ItemId) -> Self {
+        Self::new(index | Self::TREE_MASK, item)
     }
 
-    pub const fn tree(index: u16, item: u32) -> Self {
-        Self::new(index, NodeId::tree(item))
+    #[track_caller]
+    pub fn unwrap_item(&self) -> ItemId {
+        if self.index & 1 != 0 {
+            panic!("Unwrap item called on a tree node");
+        }
+        self.item
+    }
+
+    #[track_caller]
+    pub fn unwrap_tree(&self) -> ItemId {
+        if self.index & 1 != 1 {
+            panic!("Unwrap tree called on an item node");
+        }
+        self.item
+    }
+
+    pub fn to_bytes(&self) -> [u8; 5] {
+        let mut output = [0; 5];
+        output[0] = self.index;
+
+        let item_bytes = self.item.to_be_bytes();
+        debug_assert_eq!(item_bytes.len(), output.len() - 1);
```

Comment:
```suggestion
        let item_bytes = self.item.to_be_bytes();
```

> This function will panic if the two slices have different lengths.
https://doc.rust-lang.org/stable/std/primitive.slice.html#method.copy_from_slice